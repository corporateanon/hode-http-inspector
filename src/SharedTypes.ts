export type Timestamp = number;
export type Uid = string;

export interface LoggerHeaders {
    [headerName: string]: string;
}

export interface BaseLoggerRequest<T> {
    id: Uid;
    timeStart: Timestamp;
    url: string;
    body: T | null;
    headers: LoggerHeaders;
    method: string;
}

export interface BaseLoggerResponse<T> {
    timeStart: Timestamp;
    body: T | null;
    headers: LoggerHeaders;
    status: number;
}

export type LoggerRequest = BaseLoggerRequest<Buffer>;
export type LoggerResponse = BaseLoggerResponse<Buffer>;

export interface LoggerError {
    code: string;
    message: string;
    stack: string;
}

export interface LoggerEventWithError {
    request: LoggerRequest;
    response: null;
    error: LoggerError;
}

export interface LoggerEventWithResponse {
    request: LoggerRequest;
    response: LoggerResponse;
    error: null;
}

export interface ProcessData {
    pid: number;
    title: string;
    mainModule: string;
}

export type LoggerEvent = LoggerEventWithResponse | LoggerEventWithError;

export type LoggerEventHandler = (payload: LoggerEvent) => void;
export type LoggerShouldLog = (req: LoggerRequest) => boolean;
