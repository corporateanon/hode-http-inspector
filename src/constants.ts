export const WIREBIRD_DEFAULT_HOST = 'http://localhost:4380';
