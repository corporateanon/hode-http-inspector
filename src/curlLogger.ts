import GlobalHttpLogger from './GlobalHttpLogger';
import requestToCurl from './requestToCurl';

export const startCurlLogger = () => {
    const logger = new GlobalHttpLogger({
        onRequestEnd: event => {
            const curlCommand = requestToCurl(event.request);
            console.log(curlCommand);
        }
    });
    logger.start();
};
