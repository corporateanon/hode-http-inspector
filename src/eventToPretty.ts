import {
    LoggerEvent,
    LoggerRequest,
    LoggerHeaders,
    LoggerError
} from './SharedTypes';

const LINE_SEP = '------------------------------------------------';

const renderHeading = (request: LoggerRequest) =>
    `${request.method} ${request.url}`;

const renderHeaders = (headers: LoggerHeaders) =>
    Object.keys(headers)
        .map(name => `  ${name}: ${headers[name]}`)
        .join('\n');

const renderBody = (headers: LoggerHeaders, body: Buffer | null) =>
    body ? body.toString('utf8') : '<body is empty>';

const renderError = (error: LoggerError) => {
    return [
        `  Code: ${error.code}`,
        `  Message: ${error.message}`,
        `  Stack: ${error.stack}`
    ].join('\n');
};

export default function eventToPretty(event: LoggerEvent): string {
    const parts = [];
    parts.push(LINE_SEP);
    parts.push(renderHeading(event.request));
    parts.push('Request headers:');
    parts.push(renderHeaders(event.request.headers));
    parts.push('Request body:');
    parts.push(renderBody(event.request.headers, event.request.body));

    if (event.response) {
        parts.push('Response headers:');
        parts.push(renderHeaders(event.response.headers));
        parts.push('Response body:');
        parts.push(renderBody(event.response.headers, event.response.body));
    }

    if (event.error) {
        parts.push('Error:');
        parts.push(renderError(event.error));
    }
    parts.push(LINE_SEP);

    return parts.join('\n');
}
