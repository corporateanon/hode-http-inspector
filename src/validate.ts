import { SerializedLoggerEvent } from './SerializedTypes';
import { validate as schemaValidate, ValidatorResult } from 'jsonschema';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const schema = require('./schemas/SerializedLoggerEvent.json');

export const validate = (data: SerializedLoggerEvent): ValidatorResult => {
    return schemaValidate(data, schema, { nestedErrors: true });
};
