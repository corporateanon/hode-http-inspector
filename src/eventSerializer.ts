import { SerializedLoggerEvent } from './SerializedTypes';
import { LoggerEvent } from './SharedTypes';

function serializeBody<T>(data: T): T & { body?: string } {
    const _data = (data as any) as { body?: Buffer };
    if (!_data.body) {
        return data;
    }

    return {
        ...data,
        body: _data.body.toString('base64')
    };
}

export const serializeEvent = (event: LoggerEvent): SerializedLoggerEvent => {
    if (event.response) {
        return {
            request: serializeBody(event.request),
            response: serializeBody(event.response),
            error: null
        };
    }

    if (event.error) {
        return {
            request: serializeBody(event.request),
            response: null,
            error: event.error
        };
    }

    throw new Error('Cannot serialize');
};
