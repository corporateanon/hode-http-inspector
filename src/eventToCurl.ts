import { LoggerEvent } from './SharedTypes';
import escape from 'shell-escape';

interface EventToCurlOptions {
    prettyPrint?: boolean;
}

export default function eventToCurl(
    event: LoggerEvent,
    { prettyPrint = false }: EventToCurlOptions = {}
): string {
    const { request } = event;
    const lines: string[][] = [];
    lines.push(['curl', request.url]);
    lines.push(['-X', request.method]);

    for (const [key, value] of Object.entries(request.headers)) {
        lines.push(['-H', `${key}: ${value}`]);
    }

    if (request.body) {
        lines.push(['--data-binary', request.body.toString('utf8')]);
    }

    lines.push(['--compressed']);
    lines.push(['-i']);

    const separator = prettyPrint ? ' \\\n' : ' ';

    return lines.map((line) => escape(line)).join(separator);
}
