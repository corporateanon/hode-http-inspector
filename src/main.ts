import GlobalHttpLogger from './GlobalHttpLogger';
import { LoggerEvent } from './SharedTypes';

export { GlobalHttpLogger, LoggerEvent };
