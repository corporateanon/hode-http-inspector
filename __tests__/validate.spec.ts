import { serializeEvent } from '../src/eventSerializer';
import { LoggerEvent } from '../src/main';
import { ProcessData } from '../src/SharedTypes';
import { validate } from '../src/validate';

describe('validate', () => {
    const event: LoggerEvent = {
        request: {
            id     : '1',
            body   : new Buffer('hello world', 'utf8'),
            headers: {
                hello: 'world',
            },
            method   : 'GET',
            timeStart: 0,
            url      : 'https://example.com',
        },
        response: {
            body   : new Buffer('hello world', 'utf8'),
            headers: {
                hello: 'world',
            },
            rawHeaders: ['Hello', 'world'],
            status    : 200,
            timeStart : 0,
        },
        error: null,
    };

    const processData: ProcessData = {
        mainModule: 'index.js',
        pid       : 1,
        title     : 'node',
    };
    const serializedEvent = serializeEvent(event, processData);

    it('should be a correctly serialized event', () => {
        expect(serializedEvent).toMatchSnapshot();
    });

    it('should validate a correct event', () => {
        const res = validate(serializedEvent);
        expect(res.errors).toEqual([]);
        expect(res.valid).toEqual(true);
    });

    it('should validate an incorrect event', () => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const wrongEvent = serializedEvent as any;
        const res = validate({
            ...wrongEvent,
            request: { ...wrongEvent.request, url: null },
        });
        expect(
            res.errors.map((e) => ({
                message : e.message,
                property: e.property,
            }))
        ).toMatchSnapshot();
        expect(res.valid).toEqual(false);
    });
});
