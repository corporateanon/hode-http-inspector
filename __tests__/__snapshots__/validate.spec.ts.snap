// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate should be a correctly serialized event 1`] = `
Object {
  "error": null,
  "processData": Object {
    "mainModule": "index.js",
    "pid": 1,
    "title": "node",
  },
  "request": Object {
    "body": "aGVsbG8gd29ybGQ=",
    "headers": Object {
      "hello": "world",
    },
    "id": "1",
    "method": "GET",
    "remoteAddress": "127.0.0.1",
    "timeStart": 0,
    "url": "https://example.com",
  },
  "response": Object {
    "body": "aGVsbG8gd29ybGQ=",
    "headers": Object {
      "hello": "world",
    },
    "rawHeaders": Array [
      "Hello",
      "world",
    ],
    "status": 200,
    "timeStart": 0,
  },
}
`;

exports[`validate should validate an incorrect event 1`] = `
Array [
  Object {
    "message": "does not match allOf schema <#/definitions/SerializedLoggerEventWithError> with 3 error[s]:",
    "property": "instance",
  },
  Object {
    "message": "is not of a type(s) object",
    "property": "instance.error",
  },
  Object {
    "message": "is not of a type(s) string",
    "property": "instance.request.url",
  },
  Object {
    "message": "is not of a type(s) null",
    "property": "instance.response",
  },
  Object {
    "message": "does not match allOf schema <#/definitions/SerializedLoggerEventWithResponse> with 1 error[s]:",
    "property": "instance",
  },
  Object {
    "message": "is not of a type(s) string",
    "property": "instance.request.url",
  },
  Object {
    "message": "is not any of [subschema 0],[subschema 1]",
    "property": "instance",
  },
]
`;
